File: mac/Grimly/Services/FoundryLocalClient.swift
import Foundation

class FoundryLocalClient {
    private let settingsService: SettingsService
    /// Optional — used only for retry-with-endpoint-refresh. Kept optional
    /// so a smaller test harness can construct a client without a manager.
    private let foundryManager: FoundryManager?
    /// Optional external providers (Ollama, LM Studio). When the selected
    /// model carries a provider prefix ("ollama:llama3"), requests route to
    /// that provider's OpenAI-compatible endpoint instead of Foundry.
    private let externalProviders: ExternalLlmProviderService?

    init(settingsService: SettingsService,
         foundryManager: FoundryManager? = nil,
         externalProviders: ExternalLlmProviderService? = nil) {
        self.settingsService = settingsService
        self.foundryManager = foundryManager
        self.externalProviders = externalProviders
    }

    /// Prepended to every system prompt. Qwen (and to a lesser extent Phi)
    /// otherwise drifts to Chinese for anything longer than a paragraph,
    /// regardless of the input language. Anchors the reply to the input's
    /// language and calls out English explicitly for the common case.
    private static let languageAnchor =
        "Reply in the same language as the input text. If the input is in English, respond in English.\n\n"

    func getEditedText(
        originalText: String,
        mode: EditingMode,
        customPrompt: String? = nil,
        temperature: Double? = nil
    ) async throws -> String {
        let modePrompt = (mode == .customPrompt && customPrompt != nil)
            ? customPrompt!
            : mode.systemPrompt
        let systemPrompt = Self.languageAnchor + modePrompt

        // First attempt with the endpoint we have in settings.
        do {
            return try await attempt(
                systemPrompt: systemPrompt,
                originalText: originalText,
                mode: mode,
                explicitTemperature: temperature
            )
        } catch is CancellationError {
            throw CancellationError()
        } catch let urlErr as URLError {
            // Connection- or transport-layer problems only. Anything else
            // (400s, malformed JSON) rethrows immediately — retrying won't
            // help.
            let transient: Set<URLError.Code> = [
                .cannotConnectToHost, .networkConnectionLost, .notConnectedToInternet,
                .timedOut, .badServerResponse, .cannotFindHost, .dnsLookupFailed,
                .resourceUnavailable, .cannotLoadFromNetwork,
            ]
            guard transient.contains(urlErr.code) else { throw urlErr }

            // Endpoint refresh only makes sense for Foundry (external
            // providers have fixed localhost ports). Skip it when the
            // selected model is a provider model.
            if externalProviders?.matchProvider(settingsService.load().modelName) == nil,
               let mgr = foundryManager {
                let (running, liveEndpoint) = await mgr.checkServiceStatus()
                if running, let liveEndpoint {
                    var s = settingsService.load()
                    if s.foundryEndpoint != liveEndpoint {
                        s.foundryEndpoint = liveEndpoint
                        settingsService.save(s)
                    }
                }
            }

            // Retry once. If this also fails, the error surfaces to the caller.
            return try await attempt(
                systemPrompt: systemPrompt,
                originalText: originalText,
                mode: mode,
                explicitTemperature: temperature
            )
        }
    }

    /// Single request attempt. Reads the current endpoint from settings at
    /// call time so a retry after an endpoint refresh picks up the new port.
    private func attempt(
        systemPrompt: String,
        originalText: String,
        mode: EditingMode,
        explicitTemperature: Double?
    ) async throws -> String {
        let settings = settingsService.load()

        let finalTemp: Double
        if let temp = explicitTemperature {
            finalTemp = temp
        } else {
            let baseTemp = mode.baseTemperature
            let offset = (settings.creativity - 0.5) * 0.4
            finalTemp = min(max(baseTemp + offset, 0.0), 1.0)
        }

        // Route: a provider-prefixed model ("ollama:llama3") goes to that
        // provider's base URL with the prefix stripped off the model id;
        // a plain model id goes to Foundry's endpoint from settings.
        var modelId = settings.modelName
        var endpoint = settings.foundryEndpoint.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        var chatPath = "/v1/chat/completions"
        if let provider = externalProviders?.matchProvider(modelId) {
            if let colon = modelId.firstIndex(of: ":") {
                modelId = String(modelId[modelId.index(after: colon)...])
            }
            endpoint = provider.baseURL.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
            chatPath = provider.chatEndpoint
        }

        let request = ChatCompletionRequest(
            model: modelId,
            messages: [
                .system(systemPrompt),
                .user(originalText)
            ],
            temperature: finalTemp,
            maxTokens: settings.maxTokens
        )

        guard let url = URL(string: "\(endpoint)\(chatPath)") else {
            throw URLError(.badURL)
        }

        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.httpBody = try JSONEncoder().encode(request)

        let (data, response) = try await URLSession.shared.data(for: urlRequest)

        if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode != 200 {
            throw URLError(.badServerResponse)
        }

        let result = try JSONDecoder().decode(ChatCompletionResponse.self, from: data)
        let reply = result.choices.first?.message.content.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard !reply.isEmpty else { return originalText }

        // Response hygiene. Small on-device models (phi-3.5, qwen) leak
        // instructions into results four ways; scrub/reject each. A rejected
        // response returns the original text, which the popup reports as
        // "No changes suggested."
        let stripped = ResponseHygiene.stripMetaPreamble(reply)
        guard !stripped.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return originalText }
        if ResponseHygiene.looksLikePromptEcho(stripped, systemPrompt: systemPrompt) { return originalText }
        if ResponseHygiene.looksLikeInstructionParaphrase(stripped, systemPrompt: systemPrompt, originalText: originalText) { return originalText }
        if ResponseHygiene.looksLikeModelVerdict(stripped, originalText: originalText) { return originalText }
        return stripped
    }
}

/// Guards against small on-device models leaking their instructions into the
/// output. Ported 1:1 from the Windows `FoundryLocalClient` hygiene methods
/// (StripMetaPreamble / LooksLikePromptEcho / LooksLikeInstructionParaphrase
/// / LooksLikeModelVerdict). Kept as a standalone enum so the StyleHelper
/// client can reuse the exact same logic.
enum ResponseHygiene {

    // Words of 4+ letters — short function words (the/is/and) appear in
    // everything and would wash out the provenance signal.
    private static let wordRegex = try! NSRegularExpression(pattern: "[a-z']{4,}")

    /// Strips announcer preambles the model wraps around an otherwise-good
    /// rewrite — "The revised text is: :", "Here's the rewritten version:",
    /// "Sure, here is the edited text -". KEEPS the rewrite that follows.
    static func stripMetaPreamble(_ response: String) -> String {
        let trimmed = String(response.drop(while: { $0 == " " || $0 == "\t" || $0 == "\n" || $0 == "\r" }))
        let ns = trimmed as NSString
        let full = NSRange(location: 0, length: ns.length)
        var s = trimmed
        if let m = metaPreambleRegex.firstMatch(in: trimmed, range: full), m.range.location == 0 {
            s = ns.replacingCharacters(in: m.range, with: "")
        }
        // Mop up stragglers the model sometimes doubles after the preamble
        // (the "…is: :" case) plus any opening quote it added.
        return String(s.drop(while: { ":-–  \r\n\"\u{201C}".contains($0) }))
    }

    // The literal “ (U+201C) is embedded directly: in a Swift raw string
    // `\u{201C}` would NOT be interpreted, and ICU regex doesn't accept the
    // `\u{...}` escape form either.
    private static let metaPreambleRegex = try! NSRegularExpression(
        pattern: #"^["“']?\s*(?:sure[,.!]?\s*)?(?:here(?:'s|\s+is)\s+)?(?:the\s+|your\s+|a\s+)?(?:revised|rewritten|edited|updated|corrected|improved|conversational)\s+(?:text|version)(?:\s+is)?\s*[:\-–]+\s*"#,
        options: [.caseInsensitive])

    /// True when any substantive (20+ char) line of the system prompt shows
    /// up verbatim in the response — instruction text never legitimately
    /// belongs in edited user text.
    static func looksLikePromptEcho(_ response: String, systemPrompt: String) -> Bool {
        let lowerResponse = response.lowercased()
        for rawLine in systemPrompt.split(separator: "\n", omittingEmptySubsequences: false) {
            let line = rawLine.trimmingCharacters(in: .whitespaces)
            if line.count < 20 { continue }
            if lowerResponse.contains(line.lowercased()) { return true }
        }
        return false
    }

    /// Detects instruction paraphrase via vocabulary provenance: a genuine
    /// rewrite is assembled mostly from the user's words; a paraphrased
    /// rulebook is assembled from the prompt's words.
    static func looksLikeInstructionParaphrase(_ response: String, systemPrompt: String, originalText: String) -> Bool {
        let respTokens = words(in: response)
        if respTokens.count < 12 { return false }  // verdict detector owns tiny responses

        let promptSet = Set(words(in: systemPrompt))
        let textSet = Set(words(in: originalText))

        let promptHits = respTokens.filter { promptSet.contains($0) }.count
        let textHits = respTokens.filter { textSet.contains($0) }.count
        let promptFrac = Double(promptHits) / Double(respTokens.count)
        let textFrac = Double(textHits) / Double(respTokens.count)

        // Require both a strong absolute prompt share and a clear margin over
        // the text share so heavy rewrites (which add new wording) stay safe.
        return promptFrac > 0.55 && promptFrac > textFrac * 1.5
    }

    /// Detects verdict responses — the model narrating a judgment ("The text
    /// is now grammatically correct…") instead of returning the text. Both
    /// signals required: structural collapse + editing vocabulary the user's
    /// text doesn't use.
    static func looksLikeModelVerdict(_ response: String, originalText: String) -> Bool {
        let collapsed =
            (originalText.count >= 240 && response.count < originalText.count / 3)
            || response.count <= 160
        if !collapsed { return false }

        let lowerResponse = response.lowercased()
        let lowerOriginal = originalText.lowercased()
        for term in verdictVocabulary {
            if lowerResponse.contains(term) && !lowerOriginal.contains(term) { return true }
        }
        return false
    }

    /// Editing-meta vocabulary marking a response as being ABOUT the text
    /// rather than the text. Stored lowercase for case-insensitive contains.
    private static let verdictVocabulary: [String] = [
        "grammatical", "grammatically", "spelling error", "punctuation error",
        "no corrections", "no changes", "no edits", "no revisions",
        "revisions are complete", "revision is complete", "edits are complete",
        "already well-written", "already correct", "well-written and",
        "free of errors", "error-free", "the text is", "your text is",
        "text has been", "nothing to correct", "does not require any",
        "doesn't require any",
    ]

    private static func words(in text: String) -> [String] {
        let lower = text.lowercased()
        let ns = lower as NSString
        let matches = wordRegex.matches(in: lower, range: NSRange(location: 0, length: ns.length))
        return matches.map { ns.substring(with: $0.range) }
    }
}

File: src/Grimly.Core/Services/FoundryLocalClient.cs
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;
using Grimly.Models;

namespace Grimly.Services;

public interface IFoundryLocalClient
{
    Task<string> GetEditedTextAsync(string originalText, EditingMode mode, string? customPrompt, CancellationToken ct = default, double? temperature = null);
}

public sealed class FoundryLocalClient : IFoundryLocalClient
{
    private readonly HttpClient _httpClient;
    private readonly ISettingsService _settingsService;
    private readonly IExternalLlmProviderService? _externalProviders;
    private readonly IWindowsAiClient? _windowsAi;

    public FoundryLocalClient(
        HttpClient httpClient,
        ISettingsService settingsService,
        IExternalLlmProviderService? externalProviders = null,
        IWindowsAiClient? windowsAi = null)
    {
        _httpClient = httpClient;
        _settingsService = settingsService;
        _externalProviders = externalProviders;
        _windowsAi = windowsAi;
    }

    public async Task<string> GetEditedTextAsync(string originalText, EditingMode mode, string? customPrompt, CancellationToken ct = default, double? temperature = null)
    {
        var settings = _settingsService.Load();

        var systemPrompt = mode == EditingMode.CustomPrompt && customPrompt != null
            ? customPrompt
            : mode.GetSystemPrompt();

        // Language anchor. Qwen models (trained on mixed English + Chinese
        // data) sometimes drift to Chinese output when the system prompt is
        // short or the input is ambiguous. Prepending an explicit language
        // instruction pins the output to whatever the input was written in.
        // Cheap for models that would have stayed on-language anyway; the
        // safety net kicks in only when the model was going to drift.
        systemPrompt =
            "Reply in the same language as the input text. If the input is in English, respond in English.\n\n"
            + systemPrompt;

        // Windows AI (Aion Instruct) route. When the user picked the
        // "windows-ai" virtual model, skip Foundry's HTTP path entirely and
        // run the same prompt on-NPU. Aion has no system/user role split or
        // temperature control — instructions and text concatenate into one
        // prompt with a separator, matching how small instruct models are
        // trained to read them.
        if (_windowsAi is not null &&
            string.Equals(settings.ModelName, IWindowsAiClient.ModelId, StringComparison.OrdinalIgnoreCase))
        {
            // Text FIRST, instructions after. Aion (like other small
            // instruct models) parrots or paraphrases instructions far
            // more when they lead the prompt; leading with the text and
            // closing with an output-only reminder keeps its attention on
            // the material being edited.
            var aionResult = await _windowsAi.GenerateAsync(
                "--- TEXT ---\n" + originalText
                + "\n--- END TEXT ---\n\nApply these editing instructions to the text above:\n"
                + systemPrompt
                + "\n\nReply with ONLY the revised text. Never repeat, describe, or summarize these instructions.",
                ct);
            if (string.IsNullOrWhiteSpace(aionResult)) return originalText;
            var aionText = StripMetaPreamble(aionResult!);
            if (string.IsNullOrWhiteSpace(aionText)) return originalText;
            if (LooksLikePromptEcho(aionText, systemPrompt)) return originalText;
            if (LooksLikeInstructionParaphrase(aionText, systemPrompt, originalText)) return originalText;
            if (LooksLikeModelVerdict(aionText, originalText)) return originalText;
            return aionText;
        }

        // Compute temperature: mode baseline + creativity offset
        double finalTemp;
        if (temperature.HasValue)
        {
            finalTemp = temperature.Value; // explicit override (used by revision variants)
        }
        else
        {
            double baseTemp = mode.GetBaseTemperature();
            double offset = (settings.Creativity - 0.5) * 0.4; // -0.2 to +0.2
            finalTemp = Math.Clamp(baseTemp + offset, 0.0, 1.0);
        }

        var modelId = settings.ModelName;
        var endpoint = settings.FoundryEndpoint.TrimEnd('/');
        var chatPath = "/v1/chat/completions";
        var external = _externalProviders?.MatchProvider(modelId);
        if (external is not null)
        {
            var colon = modelId.IndexOf(':');
            modelId = colon >= 0 ? modelId[(colon + 1)..] : modelId;
            endpoint = external.BaseUrl.TrimEnd('/');
            chatPath = external.ChatEndpoint;
        }

        var request = new ChatCompletionRequest
        {
            Model = modelId,
            Temperature = finalTemp,
            MaxTokens = settings.MaxTokens,
            Messages =
            [
                ChatMessage.System(systemPrompt),
                ChatMessage.User(originalText)
            ]
        };

        var json = JsonSerializer.Serialize(request);
        var content = new StringContent(json, Encoding.UTF8, "application/json");

        // 2-minute timeout — local LLMs can be slow on long text but shouldn't take longer
        using var timeoutCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
        timeoutCts.CancelAfter(TimeSpan.FromMinutes(2));

        try
        {
            var response = await _httpClient.PostAsync(
                $"{endpoint}{chatPath}", content, timeoutCts.Token);

            if (!response.IsSuccessStatusCode)
            {
                // Surface Foundry's actual error text so the pipeline can log/
                // report something useful instead of a bare status code.
                string body;
                try { body = await response.Content.ReadAsStringAsync(timeoutCts.Token); }
                catch { body = ""; }
                throw new HttpRequestException(
                    $"Foundry returned {(int)response.StatusCode} {response.StatusCode}: {TrimBody(body)}",
                    inner: null,
                    statusCode: response.StatusCode);
            }

            var responseJson = await response.Content.ReadAsStringAsync(timeoutCts.Token);
            var result = JsonSerializer.Deserialize<ChatCompletionResponse>(responseJson);

            var reply = result?.Choices.FirstOrDefault()?.Message.Content?.Trim();
            if (string.IsNullOrWhiteSpace(reply)) return originalText;
            var replyText = StripMetaPreamble(reply!);
            if (string.IsNullOrWhiteSpace(replyText)) return originalText;
            if (LooksLikePromptEcho(replyText, systemPrompt)) return originalText;
            if (LooksLikeInstructionParaphrase(replyText, systemPrompt, originalText)) return originalText;
            if (LooksLikeModelVerdict(replyText, originalText)) return originalText;
            return replyText;
        }
        catch (OperationCanceledException) when (!ct.IsCancellationRequested)
        {
            // Our timeout fired, not the caller's cancellation
            throw new HttpRequestException("Request timed out after 2 minutes. The model may be overloaded or unresponsive.");
        }
    }

    /// <summary>
    /// Dynamic prompt-echo detector. Small on-device models sometimes
    /// return the instructions instead of (or wrapped around) the edited
    /// text. Rather than maintaining hand-curated signature lists per
    /// prompt, this checks whether any substantive line of the system
    /// prompt appears verbatim in the response — instruction lines never
    /// legitimately belong in edited user text. Also catches the common
    /// meta-commentary preambles.
    /// </summary>
    internal static bool LooksLikePromptEcho(string response, string systemPrompt)
    {
        foreach (var rawLine in systemPrompt.Split('\n'))
        {
            var line = rawLine.Trim();
            // Short lines ("RULES:", "1.") are too generic to be reliable
            // evidence; 20+ chars of verbatim instruction text is.
            if (line.Length < 20) continue;
            if (response.Contains(line, StringComparison.OrdinalIgnoreCase))
                return true;
        }

        return false;
    }

    /// <summary>
    /// Detects instruction paraphrase — the model restating the prompt's
    /// rules in its own words ("It avoids complex vocabulary. It uses
    /// contractions…"), which verbatim line-matching can't catch. The tell
    /// is vocabulary provenance: a genuine rewrite is assembled mostly from
    /// the user's own words, while a paraphrased rulebook is assembled from
    /// the prompt's words. Flag responses whose word stock leans clearly
    /// toward the prompt.
    /// </summary>
    internal static bool LooksLikeInstructionParaphrase(string response, string systemPrompt, string originalText)
    {
        var respTokens = WordRegex.Matches(response.ToLowerInvariant())
            .Select(m => m.Value).ToList();
        // Tiny responses don't give the statistic enough to grip on; the
        // verdict detector owns that range.
        if (respTokens.Count < 12) return false;

        var promptSet = new HashSet<string>(
            WordRegex.Matches(systemPrompt.ToLowerInvariant()).Select(m => m.Value));
        var textSet = new HashSet<string>(
            WordRegex.Matches(originalText.ToLowerInvariant()).Select(m => m.Value));

        double promptFrac = respTokens.Count(promptSet.Contains) / (double)respTokens.Count;
        double textFrac = respTokens.Count(textSet.Contains) / (double)respTokens.Count;

        // Rewrites typically score textFrac ≥ 0.6 (the content is the
        // user's). Paraphrased instructions score promptFrac high and
        // textFrac low. Require both a strong absolute prompt share and a
        // clear margin over the text share so heavy rewrites (which add
        // some new wording) stay safe.
        return promptFrac > 0.55 && promptFrac > textFrac * 1.5;
    }

    // Words of 4+ letters — short function words (the/is/and) appear in
    // everything and would wash out the signal.
    private static readonly Regex WordRegex = new(@"[a-z']{4,}", RegexOptions.Compiled);

    /// <summary>
    /// Strips announcer preambles the model wraps around an otherwise-good
    /// rewrite — "The revised text is: :", "Here's the rewritten version:",
    /// "Sure, here is the edited text -" and friends. Unlike echo/verdict
    /// detection this KEEPS the response: the rewrite that follows the
    /// preamble is usually exactly what the user asked for.
    /// </summary>
    internal static string StripMetaPreamble(string response)
    {
        var s = MetaPreambleRegex.Replace(response.TrimStart(), "", 1);
        // Mop up stragglers the model sometimes doubles after the preamble
        // (the "…is: :" case) plus any opening quote it added.
        return s.TrimStart(':', '-', ' ', '\r', '\n', '"', '“');
    }

    private static readonly Regex MetaPreambleRegex = new(
        @"^[""“']?\s*(?:sure[,.!]?\s*)?(?:here(?:'s|\s+is)\s+)?(?:the\s+|your\s+|a\s+)?(?:revised|rewritten|edited|updated|corrected|improved|conversational)\s+(?:text|version)(?:\s+is)?\s*[:\-–]+\s*",
        RegexOptions.Compiled | RegexOptions.IgnoreCase);

    /// <summary>
    /// Detects verdict responses — the model narrating a judgment about the
    /// text ("The revisions are complete. The text is now grammatically
    /// correct…") instead of returning the text. Two signals, both
    /// required:
    ///   1. Structural collapse — the response is far too short to be a
    ///      rewrite of the input. Even aggressive modes (Bullet Points,
    ///      Cut Filler) keep most of the content; verdicts are a sentence
    ///      or two regardless of input size.
    ///   2. Editing vocabulary — the response talks about grammar/spelling/
    ///      revisions in words the user's own text doesn't use.
    /// Callers map a verdict to "no changes" by returning the original
    /// text, which the popup reports as "No changes suggested."
    /// </summary>
    internal static bool LooksLikeModelVerdict(string response, string originalText)
    {
        // Structural gate. Short inputs can't distinguish a verdict from a
        // legitimate tightening, so only fire when the input is substantial
        // and the response collapsed to a fraction of it (or is just tiny).
        bool collapsed =
            (originalText.Length >= 240 && response.Length < originalText.Length / 3)
            || response.Length <= 160;
        if (!collapsed) return false;

        foreach (var term in VerdictVocabulary)
        {
            if (response.Contains(term, StringComparison.OrdinalIgnoreCase) &&
                !originalText.Contains(term, StringComparison.OrdinalIgnoreCase))
                return true;
        }
        return false;
    }

    /// <summary>Editing-meta vocabulary that marks a response as being
    /// ABOUT the text rather than the text. Stems, so tense/phrasing
    /// variants ("is/now grammatically correct", "revision/revisions")
    /// all match.</summary>
    private static readonly string[] VerdictVocabulary =
    {
        "grammatical", "grammatically", "spelling error", "punctuation error",
        "no corrections", "no changes", "no edits", "no revisions",
        "revisions are complete", "revision is complete", "edits are complete",
        "already well-written", "already correct", "well-written and",
        "free of errors", "error-free", "the text is", "your text is",
        "text has been", "nothing to correct", "does not require any",
        "doesn't require any",
    };

    private static string TrimBody(string body)
    {
        if (string.IsNullOrWhiteSpace(body)) return "(no body)";
        // Response bodies can be large; keep a reasonable preview.
        var oneLine = body.Replace('\n', ' ').Replace('\r', ' ').Trim();
        return oneLine.Length > 300 ? oneLine[..300] + "…" : oneLine;
    }
}
